File: 프로그래머스/~2021/베스트앨범.java
import java.util.*;
class Solution {
    HashMap<String, PriorityQueue<Node>> map = new HashMap<>();
    HashMap<String, Integer> map2 = new HashMap<>();
    class Node implements Comparable<Node> {
        int idx, n;
        public Node(int idx, int n){
            this.idx = idx;
            this.n = n;
        }
        public int compareTo(Node node){
            return node.n - n;
        }
    }
    
    public int[] solution(String[] genres, int[] plays) {
        for(int i = 0; i < genres.length; i++){
            String gen = genres[i];
            int pl = plays[i];
            
            PriorityQueue<Node> pq = map.get(gen);
            if(pq == null){
                pq = new PriorityQueue<>();
            }
            pq.add(new Node(i, pl));
            map.put(gen, pq);
            
            map2.put(gen, map2.getOrDefault(gen, 0) + pl);
        }
        
        List<Map.Entry<String, Integer>> list = new LinkedList<>(map2.entrySet());
        Collections.sort(list, new Comparator<Map.Entry<String, Integer>>(){
            public int compare(Map.Entry<String, Integer> a, Map.Entry<String, Integer> b){
                return b.getValue() - a.getValue();
            }
        });
        
        List<Integer> re = new ArrayList<>();
        for(Map.Entry<String, Integer> en : list){
            String gen = en.getKey();
            PriorityQueue<Node> pq = map.get(gen);
            Node first = pq.poll();
            re.add(first.idx);
            
            if(pq.isEmpty())
                continue;
            
            Node second = pq.poll();
            re.add(second.idx);
        }
        
        int[] reArr = new int[re.size()];
        int arrIdx = 0;
        for(int idx : re){
            reArr[arrIdx++] = idx;
        }
        
        return reArr;
    }
}

File: 프로그래머스/~2021/베스트앨범.js
function solution(genres, plays) {
    var map = {};
    for(var i = 0; i < genres.length; i++){
        if(map[genres[i]] === undefined)
            map[genres[i]] = [];
        map[genres[i]].push([i, plays[i]]);
        // map[genres[i]] = map[genres[i]] ? map[genres[i]].push([i, plays[i]]) : [[i, plays[i]]];
    }

    var mostPlayOfGenre = [];
    for(var k in map){
        var v = map[k];
        var sum = 0;
        for(var j in v){
            sum += v[j][1];
        }
        v.sort((a, b) => b[1] === a[1] ? a[0] - b[0] : b[1] - a[1]);
        var temp = [sum, v[0]];
        if(v.length > 1){
            temp.push(v[1]);
        }
        mostPlayOfGenre.push(temp);
    }

    mostPlayOfGenre.sort((a, b) => b[0] - a[0]);
    var re = [];
    for(var i in mostPlayOfGenre){
        var songs = mostPlayOfGenre[i];
        for(var j = 1; j < songs.length; j++){
            var song = songs[j];
            re.push(song[0]);
        }
    }
    return re;
}